library("sleuth")

# input file generated by:
# for i in fastqfiles.gz; do kallisto quant --single -i ~/Data/genomes/kallisto/hg19-cnda.ncrna.idx -o
# /media/snfrbert/Ayana/5.RNA-Seq/all/${i%.fastq.gz}/kallisto -b 25 -l 101 -s 1 -t 8 $i; done

mart <- biomaRt::useMart(biomart = "ENSEMBL_MART_ENSEMBL", dataset = "hsapiens_gene_ensembl", host = 'grch37.ensembl.org')
t2g <- biomaRt::getBM(attributes = c("ensembl_transcript_id", "ensembl_gene_id", "external_gene_name"), mart = mart)
t2g <- dplyr::rename(t2g, target_id = ensembl_transcript_id, ens_gene = ensembl_gene_id, ext_gene = external_gene_name)

base_dir <- "/media/snfrbert/Ayana/5.RNA-Seq/all"
samp <- "op120" 
information <- "op120-info.txt" 
# Info file must be sorted like files in directory!
# Also control and input-condition should be named, such that control comes first in alphabet
# otherwise test in the wrong way!

sample_id <- dir(file.path(base_dir, samp)) 
kallisto_dirs <- sapply(sample_id, function(id) file.path(base_dir,samp,id,"kallisto"))
s2c <- read.table(file.path(base_dir,information), header = TRUE, stringsAsFactors=FALSE)
s2c <- dplyr::select(s2c, sample = sample, individuum, lane, condition)
s2c <- dplyr::mutate(s2c, path = kallisto_dirs)


# reads in aggregating by gene, just remove option for transcript reading
so <- sleuth_prep(s2c, ~ condition, target_mapping = t2g, aggregation_column = 'ext_gene')

so <- sleuth_fit(so)
so <- sleuth_wt(so, which_beta = 'conditiondmog')
sleuth_live(so)

# may be save for later use, e.g. without internet connection
saveRDS(so, file = 'op131-so.rds')



library("sleuth")

# input file generated by:
# for i in fastqfiles.gz; do kallisto quant --single -i ~/Data/genomes/kallisto/hg19-cnda.ncrna.idx -o
# /media/snfrbert/Ayana/5.RNA-Seq/all/${i%.fastq.gz}/kallisto -b 25 -l 101 -s 1 -t 8 $i; done

mart <- biomaRt::useMart(biomart = "ENSEMBL_MART_ENSEMBL", dataset = "hsapiens_gene_ensembl", host = 'grch38.ensembl.org')
t2g <- biomaRt::getBM(attributes = c("ensembl_transcript_id", "ensembl_gene_id", "external_gene_name"), mart = mart)
t2g <- dplyr::rename(t2g, target_id = ensembl_transcript_id, ens_gene = ensembl_gene_id, ext_gene = external_gene_name)

base_dir <- "/media/snfrbert/Ayana/5.RNA-Seq/all"
samp <- "whole" 
information <- "whole-info.txt" 
# Info file must be sorted like files in directory!
# Also control and input-condition should be named, such that control comes first in alphabet
# otherwise test in the wrong way!

sample_id <- dir(file.path(base_dir, samp)) 
kallisto_dirs <- sapply(sample_id, function(id) file.path(base_dir,samp,id,"kallisto"))
s2c <- read.table(file.path(base_dir,information), header = TRUE, stringsAsFactors=FALSE)
s2c <- dplyr::select(s2c, sample = sample, individuum, condition)
s2c <- dplyr::mutate(s2c, path = kallisto_dirs)


# reads in aggregating by gene, just remove option for transcript reading
so <- sleuth_prep(s2c, ~ condition, target_mapping = t2g), aggregation_column = 'ext_gene')

so <- sleuth_fit(so)
so <- sleuth_wt(so, which_beta = 'conditiondmog')
sleuth_live(so)

# may be save for later use, e.g. without internet connection
saveRDS(so, file = 'whole-so.rds')
